File: src/main/scala/com/ons/sml/businessMethods/methods/Apportionment.scala
package com.ons.sml.businessMethods.methods

import com.ons.sml.businessMethods.impl.ApportionmentImpl
import org.apache.spark.sql.{DataFrame, Dataset, Row}


class Apportionment (val dfIn: DataFrame){

  if (dfIn == null ) throw new Exception("DataFrame cannot be null")

  import ApportionmentImpl._

  val defaultCol = "AppResult"

  private def mandatoryArgCheck(arg1 : String, arg2 : String) : Unit = {

    if ((arg1 == null) || (arg2 == null)) throw new Exception("Missing mandatory argument")
  }

  def app1(df: DataFrame, groupByCol : String, aggCol : String, appCol : String = defaultCol) : DataFrame = {

    mandatoryArgCheck(groupByCol, aggCol)

    val dF = if (df == null) dfIn else df

    dF.checkColNames(Seq(groupByCol, aggCol))
      .app1(groupByCol, aggCol, appCol)
  }


  def app2(df: DataFrame, groupByCol : String, aggCol : String, appCol : String = defaultCol) : DataFrame = {

    mandatoryArgCheck(groupByCol, aggCol)

    val dF = if (df == null) dfIn else df

    dF.checkColNames(Seq(groupByCol, aggCol))
      .app2(groupByCol, aggCol, appCol)
  }

  def app3(df: DataFrame, groupByCol : String, aggCol : String, appCol : String = defaultCol) : DataFrame = {

    mandatoryArgCheck(groupByCol, aggCol)

    val dF = if (df == null) dfIn else df

    dF.checkColNames(Seq(groupByCol, aggCol))
      .app3(groupByCol, aggCol, appCol)
  }

  def app4(df: DataFrame, groupByCol : String, aggCol : String, appCol : String = defaultCol) : DataFrame = {

    mandatoryArgCheck(groupByCol, aggCol)

    val dF = if (df == null) dfIn else df

    dF.checkColNames(Seq(groupByCol, aggCol))
      .app4(groupByCol, aggCol, appCol)
  }
}

object Apportionment {

  def apportionment(df: DataFrame) : Apportionment = new Apportionment(df)
}

File: python/sparkts/onsMethods/Apportionment.py
from pyspark.sql import DataFrame


class Apportionment():

    defaultCol = 'AppResult'

    def __init__(self, df = None):

        if df is None : raise TypeError

        self._df = df

        self._jApp = self._df._sc._jvm.com.ons.sml.api.java.methods.JavaApportionmentFactory.apportionment(self._df._jdf)


    def __mandatoryArgumentCheck(self, arg1, arg2):

        if ((arg1 is None) or (arg2 is None)) : raise TypeError

    def app1(self, df = None, groupByCol = None, aggCol = None, appColName = defaultCol):

        self.__mandatoryArgumentCheck(groupByCol, aggCol)

        if df is None: df = self._df

        return DataFrame(self._jApp.app1(df._jdf, groupByCol, aggCol, appColName),  df.sql_ctx)

    def app2(self, df = None, groupByCol = None, aggCol = None, appColName = defaultCol):

        self.__mandatoryArgumentCheck(groupByCol, aggCol)

        if df is None: df = self._df

        return DataFrame(self._jApp.app2(df._jdf, groupByCol, aggCol, appColName),  df.sql_ctx)

    def app3(self, df = None, groupByCol = None, aggCol = None, appColName = defaultCol):

        self.__mandatoryArgumentCheck(groupByCol, aggCol)

        if df is None: df = self._df

        return DataFrame(self._jApp.app3(df._jdf, groupByCol, aggCol, appColName),  df.sql_ctx)

    def app4(self, df = None, groupByCol = None, aggCol = None, appColName = defaultCol):

        self.__mandatoryArgumentCheck(groupByCol, aggCol)

        if df is None: df = self._df

        return DataFrame(self._jApp.app4(df._jdf, groupByCol, aggCol, appColName),  df.sql_ctx)


def apportionment(df) :

    return Apportionment(df)